### ISSUE TITLE
Memory use is ~2x expected

### ISSUE THREAD

**Issue by johnlees** [tseemann/snippy#290]
With snippy 4.3.3:
```
snippy --outdir 6710_6_17 --cpus 1 --ram 2 --cleanup --reference D39V.gb --pe1 6710_6#17_1.fastq.gz --pe2 6710_6#17_2.fastq.gz
```
This is a *S. pneumoniae* genome with ~1Gb of sequence in the reads. Although I have specified 2Gb of RAM in the command the max used was 4.1Gb.

I think this occurs when the following command is run
```
bwa mem  -Y -M -R '@RG\tID:6710_6_17\tSM:6710_6_17' -t 1 reference/ref.fa /lustre/scratch118/infgen/team81/jl11/pdoc/transmissibility/D39V_map/pf_lanes/6710_6#17_1.fastq.gz /lustre/scratch118/infgen/team81/jl11/pdoc/transmissibility/D39V_map/pf_lanes/6710_6#17_2.fastq.gz | samclip --max 10 --ref reference/ref.fa.fai | samtools sort -n -l 0 -T /tmp/snippy.1057. --threads 1 -m 2000M | samtools fixmate -m - - | samtools sort -l 0 -T /tmp/snippy.1057. --threads 1 -m 2000M | samtools markdup -T /tmp/snippy.1057. -r -s - - > snps.bam 2>> snps.log
```

A relevant message from this phase:
```
[bam_sort_core] merging from 1 files and 1 in-memory blocks...
[E::hts_open_format] Failed to open file /tmp/snippy.1057..0000.bam
samtools sort: failed to create temporary file "/tmp/snippy.1057..0000.bam": File exists
```

I also tried setting `--tmpdir` to the cwd (which definitely has sufficient space), which did not fix this issue.

My guess is that as this uses pipes, the two sort commands are run concurrently as data is available, each taking up to `-m` memory. I tracked the memory use over the run and it starts to exceed the threshold when the `bwa mem` and first `samtools sort` are both running, then reaches double during the sort > fixmate > sort phase, after `bwa mem` has finished.
I think this is also leading to the above error/warning, as eventually two sort commands are running concurrently and trying to use the same tmp files.

A solution could be to halve the input value from `--ram` to give the `-m` for these commands in the pipe, and give each one separate `-T` arguments.

**Comment by tseemann** [comment]
Thanks for this double bug report. 

(1) You are spot on, I use the RAM amount twice in a pipe so i need to halve it.

(2) You are right about the temp clash too. That is bad! I have never encountered it because my sorting is always done in RAM. But I will change it to use the `-T folder` syntax which adds random strings into the temp file (see https://github.com/samtools/samtools/issues/432#issuecomment-484100434)

**Comment by tseemann** [comment]
@johnlees i am unsure how to control the RAM usage of `bwa mem` and `samtools markdup` however.

BWA is affected by the index, but also the chunking, so this might help:
```
       -K INT        process INT input bases in each batch regardless of nThreads (for reproducibility) []
```

**Comment by tseemann** [comment]
I've released tag 4.4.0 now.

**Comment by johnlees** [comment]
Cool, thanks for fixing so quickly! 
I think the main process sending stuff into memory at this kind of threshold is sort, so I would think that markdup and bwa mem won't cause an issue (based on my limited testing)
